monai/metrics/metric.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+import torch
+
+from monai.config import TensorOrList
+from monai.utils import evenly_divisible_all_gather
+
+
+class Metric(ABC):
```

Comment:
thanks, I rearranged a few APIs based on your impl. https://gist.github.com/wyli/0733d7965412092244e08f237bbb8205 is this feasible? in this case we'll have iterationMetric for the low-level impl. and cumulative interface might be useful for some of the metrics

Reply:
Hi @wyli ,

Thanks very much for your review and suggestions, let me re-think it and try to adjust it again tomorrow.

Reply:
BTW, I checked your sample code, may I know what's the difference between `IterationMetric` and `EpochMetric`? In current PR, the difference is the `y_pred` vs `scores`.

Thanks.

Reply:
I amended the script, `EpochMetric` becomes a generic wrapper of IterationMetric, but ROCAUC is not compatible with the wrapper, because it aggregate differently

Reply:
Hi @wyli ,

I checked your latest script, making `Cumulative` a separate class is a good idea. But I think making `EpochMetric` a wrapper is not very clear, every time users want to use MeanDice, do they need to create both MeanDiceMetric and this wrapper? And maybe different metrics have different aggregation methods(take confusion_matrix as example), we can't use the same wrapper?
Maybe we should not distinguish MeanDice and AUC, below script is slightly changed from your proposal and tried to define unified base classes for all metrics:
```py
class Metric:
  def __call__(self, ...):
    # iteration-level compute

class Cumulative:
    # utility class for cumulative metrics
  def __init__(self, buffer_count=1)
    # initialize metric buffer
    self._buffers = [[] for _ in buffer_count]
    self._synced_tensors = [torch.tensor(0) for _ in buffer_count]
    self.buffer_count = buffer_count

  def add(self, *data):
      for i, d in enumerate(data):
          self._buffers[i].append(d)

  def reset(self, ...):
    # reset the metric buffer
    self._buffers = [[] for _ in self.buffer_count]
    self._synced_tensors = [torch.tensor(0) for _ in self.buffer_count]

  def aggregate(self, ...):
    # compute result based on buffer (build by multiple __call__)
    self._sync()
    
  def _sync(self, ...):
    # sync up distributed metric when aggregate
    self._synced_tensors = [evenly_divisible_all_gather(data=torch.cat(b, dim=0), concat=True) for b in self._buffers]

  def get_synced_tensors(self):
    # need to get raw metric details to generate reports
    return self._synced_tensors

### concrete classes ###
class CumulativeMetric(Metric, Cumulative):
  def __call__(self, ...):
    # compute based on _compute_*
    if isinstance(list):
      ret = self._compute_list()
    else:
      ret = _compute_tensor()
    self.add(ret)

  def _compute_list(self, ...):
    # per iteration results on list data

  def _compute_tensor(self, ...):
    # per iteration results on tensor data

class ROCAUCMetric(CumulativeMetric):
  def __init__(self, ...):
      super().__init__(buffer_count=2)

  def _compute_tensor(self, ...):

  def aggregate(self, ...):
      super().aggregate()
      y_pred, y = self._synced_tensors[0], self._synced_tensors[1]
      ...

class MeanDiceMetric(CumulativeMetric):
  def __init__(self, ...):
      super().__init__(buffer_count=1)

  def _compute_tensor(self, ...):

  def aggregate(self, ...):
      super().aggregate()
      scores = self._synced_tensors[0]
      ...
```
What do you think?

Thanks.

Reply:
sure I updated the script to unify the parts

Reply:
Cool, I think we almost aligned on the APIs of metrics, I will update the PR today based on our discussion.

Thanks.

Reply:
Updated the base classes in the latest commit.
Thanks for the discussion.

Reply:
thanks the base classes look good to me, I'll review/revise the details if you fix the tests

Reply:
Sure, I will try to update all the tests ASAP.

Thanks.